### PR TITLE
[Final] Isaac's Optimal Font Size Stuff So Far

### PR DESCRIPTION
**Pull request by aab29**
I finally figured out how to isolate a range of commits so that we can use just his code--even though he's starting from a different base-branch.  For future reference, use "git cherry-pick" with a range of commits.  :octocat:

We're really the first ones to test this code, and it's still far from perfect.  But it's good enough that we can use it to find the right font size to fit text into reasonably-sized boxes... with a little fudging on the Lua side.  Next on the list is getting it to work with `setGlyphScale()` so that we can do non-blurry text on devices with a `contentScale` greater than 1.0.  :eyes: 

If you merge https://github.com/mindsnacks/Ahu/pull/54 into Ahu, you can actually use this stuff.  :smile:

### CODE REVIEW

File: src/moaicore/MOAIFont.cpp [hunk 1]
```diff
@@ -485,6 +533,206 @@ bool MOAIFont::IsWhitespace ( u32 c ) {
...
+			wRatio = width / boxWidth;
+			hRatio = height / boxHeight;
+			float minRatio = (wRatio < hRatio) ? wRatio : hRatio;
+			
+			optimumSize = minRatio * optimumSize;
+		}
+		else {
+			// else, try finding a new calculated size that fits
+			
+			// start at calcSize and go down by 1% or one font size, whichever is greater
```
Comment (aab29) [inline src/moaicore/MOAIFont.cpp:667]: We'll probably eventually need to change this to a binary search for performance reasons... but no sense optimizing prematurely.  :open_hands:

Comment (JaviSoto) [inline src/moaicore/MOAIFont.cpp:667]: Yeah, this bit us in the app :P

File: src/moaicore/MOAIFont.cpp [hunk 2]
```diff
@@ -485,6 +533,206 @@ bool MOAIFont::IsWhitespace ( u32 c ) {
...
+				textBox -> ScheduleLayout();
+				
+				// find out if last character renders
+				lastCharacterDidRender = textBox->GetBoundsForRange(textLength - 1, 1, testRect);
+				if (lastCharacterDidRender) {
+					break;
+				}
+				
+				// reduce size and try again
+				testSize *= 0.99f;
```
Comment (jeiting) [inline src/moaicore/MOAIFont.cpp:689]: This actually produces a natural decay (e^-x) of the text size, not a linear.

Comment (aab29) [inline src/moaicore/MOAIFont.cpp:689]: Whoah, yeah.  I wonder if we should be commenting directly on Isaac's repository for some of this stuff?  :question: 

https://github.com/IkeBart/moai-dev/blob/02146081059d9bfe211a0f25c6642edf4c10ae80/src/moaicore/MOAIFont.cpp

**Pull request merged**